src/UI/Tabs.re

```diff
@@ -10,45 +10,171 @@ open Rench;
 let noop = () => ();
 
 type tabInfo = {
+  editorId: int,
   title: string,
-  active: bool,
   modified: bool,
-  onClick: Tab.tabAction,
-  onClose: Tab.tabAction,
 };
 
 let component = React.component("Tabs");
 
-let toTab = (theme, mode, uiFont, numberOfTabs, active, index, t: tabInfo) =>
+let toTab = (theme, mode, uiFont, numberOfTabs, active, activeEditorId, index, t: tabInfo) =>
   <Tab
     theme
     tabPosition={index + 1}
     numberOfTabs
     title={Path.filename(t.title)}
-    active={t.active}
+    active={Some(t.editorId) == activeEditorId}
     showHighlight=active
     modified={t.modified}
     uiFont
     mode
-    onClick={t.onClick}
-    onClose={t.onClose}
+    onClick={() => GlobalContext.current().openEditorById(t.editorId)}
+    onClose={() => GlobalContext.current().closeEditorById(t.editorId)}
   />;
 
-let viewStyle = Style.[flexDirection(`Row)];
+let measureWidth: option(node) => int = fun
+  | Some(outer) => outer#measurements().width
+  | None => 0;
+
+let measureOverflow: option(node) => int = fun
+  | Some(outer) => {
+      let inner = outer#firstChild();
+      max(0, inner#measurements().width - outer#measurements().width);
+    }
+  | None => 0;
+
+let measureChildOffset: int => option(node) => option((int, int)) = index => fun
+  | Some(outer) => {
+      let rec loop = (i, offset) => fun
+        | [] =>
+          None
+
+        | [(child: node), ..._] when i == index =>
+          Some((offset, child#measurements().width))
+
+        | [(child: node), ...rest] =>
+          loop(i + 1, offset + child#measurements().width, rest);
+
+      loop(0, 0, outer#firstChild()#getChildren());
+    }
+  | None => None;
+
+let findIndex = (predicate, list) => {
+  let rec loop = i => fun
+    | [] => None
+    | [head, ...tail] when predicate(head) => Some(i)
+    | [_, ...tail] => loop(i + 1, tail);
+  loop(0, list);
+};
+
+let isPendingRender: option(node) => bool = fun
+  | Some(outer) => outer#firstChild()#firstChild()#measurements().width < 0
```

Comment:
Wow! Nice find on how to figure out if a render is pending. 

I wonder if we could provide a helper API on the Revery side to make this more ergonomic?

Reply:
I hope so!

This is basically just what i tried to do, and found to return garbage data in some circumstances. It worries me a bit that such a widely used API can just return garbage like that though. It kind of makes the point of using a safe language a bit moot. Should it be considered a bug in revery?

The easy fix is probably just to have measurements return an `option`, have it detect `MIN_INT` on some measure and return `None` then. but that will still require this `onDimensionsChange` hack everywhere this can happen. Perhaps we could have measurements take a callback instead, that's called when rendering is done?

Reply:
>  It worries me a bit that such a widely used API can just return garbage like that. It kind of makes the point of using a safe language a bit moot. Should it be considered a bug in revery?

Agree 100% ! Good observation. The garbage is a problem - we should be leveraging the language to make it safe and easy to create UI (like we do for things like hooks).

>  Perhaps we could have measurements() take a callback instead, that's called when rendering is done?

I like this idea!